gulp-core-build/src/jsonUtilities/SchemaValidator.ts

```diff
@@ -0,0 +1,89 @@
+import * as os from 'os';
+import * as fs from 'fs';
+import * as path from 'path';
+
+import Validator = require('z-schema');
+import jju = require('jju');
+
+const schemaKey: string = '$schema';
+
+export interface ISchemaValidatorResult {
+  details?: ZSchema.SchemaError[];
+  name?: string;
+  message?: string;
+}
+
+/**
+ * Wrapper functions around z-schema which help improve ease of use
+ */
+export class SchemaValidator {
+  private static _schemaValidator: ZSchema.Validator = new Validator({
+    breakOnFirstError: true,
+    noExtraKeywords: true,
+    noTypeless: true
+  });
+
+  /**
+   * A function which validates a dataFile against a schemFile, both specified
+   * as paths. It will throw if there is an issue with the file,
+   * otherwise it will return the validated datafile.
+   */
+  public static readAndValidateJson<TResult>(dataFilePath: string, schemaFilePath: string): TResult {
+    const data: TResult = this.readCommentedJsonFile(dataFilePath) as TResult;
+    const schema: Object = this.readCommentedJsonFile(schemaFilePath);
+    this.validate(data, schema, dataFilePath);
+    return data;
+  }
+
+  /**
+   * A function which validates a data object against a schema object.
+   * It will throw if there is an issue with the data object.
+   * For an improved error message, pass the filename in the optional third parameter
+   */
+  public static validate(data: {}, schema: {}, dataFilePath?: string): void {
+    if (!this._schemaValidator.validate(data, schema)) {
+      const error: ISchemaValidatorResult = this._schemaValidator.getLastError();
+      throw this.getFormattedErrorMessage(error, dataFilePath);
+    }
+    return undefined;
+  }
+
+  public static getFormattedErrorMessage(error: ISchemaValidatorResult, dataFilePath?: string): string {
+    const errorMessage: string =
+      (dataFilePath ? `Error parsing file '${path.basename(dataFilePath)}'${os.EOL}` : '') +
+      this._extractInnerErrorMessages(error.details).join(os.EOL);
+
+    return os.EOL + 'ERROR: ' + errorMessage + os.EOL + os.EOL;
+  }
+
+  public static readCommentedJsonFile<TResult>(filename: string): TResult {
+    const contents: Buffer = fs.readFileSync(filename);
```

Comment:
Just be conventional.  Configuration goes in data files.  Config files have comments.  If you make a mistake in your config file, you get a configuration error and not a general compiler error.  If you are doing something exotic, you can write free-form program code, but it doesn't go in the config file.  This basic model has a long and very successful history.